File: docs/JAMIA/comparisons/RPython/CDA/22591174_and_or_filter_confusion.ts
/**
 * RPython SO#22591174 — pandas: multiple conditions while indexing - unexpected behavior
 * Effect: IF (silent incorrect functionality)
 * Bug class: Operator overloading
 *
 * In pandas, boolean indexing with | and & on negated conditions confuses users.
 * df[(df.a != -1) | (df.b != -1)] keeps rows where EITHER is != -1, but users
 * expect it to drop rows where EITHER equals -1.
 *
 * In tidy-ts, filter() takes a predicate function using standard JS logical
 * operators (&&, ||). The semantics are "keep rows where predicate returns true."
 * There is no operator overloading confusion. The bug is structurally absent —
 * this is a language semantics issue, not a type issue. No @ts-expect-error
 * is applicable because the fix is clearer operator semantics, not type checking.
 */
import { createDataFrame } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { a: 0, b: 0 },
  { a: -1, b: -1 },
  { a: 2, b: 2 },
  { a: -1, b: 3 },
  { a: 4, b: -1 },
]);

// "keep rows where a != -1 AND b != -1" — && means what you expect
const filtered = df.filter((r) => r.a !== -1 && r.b !== -1);
filtered.print();

File: docs/JAMIA/comparisons/RPython/CDA/22591174_and_or_filter_confusion.py
"""
RPython SO#22591174 — pandas: multiple conditions while indexing - unexpected behavior
Effect: IF (silent incorrect functionality)

Users expect `|` to mean "drop rows matching ANY condition" but pandas boolean
indexing with `|` keeps rows matching ANY condition. The operator semantics are
correct but confusing — no error when logic is inverted.
"""
import pandas as pd

df = pd.DataFrame({'a': [0, -1, 2, -1, 4], 'b': [0, -1, 2, 3, -1]})

print("Original:")
print(df)
print()

# User wants to drop rows where EITHER column is -1
# BUG: uses | thinking "drop if a == -1 OR b == -1"
# Actually keeps rows where a != -1 OR b != -1 (i.e., only drops if BOTH are -1)
df_wrong = df[(df.a != -1) | (df.b != -1)]

print("Using | (user thinks: drop any -1, actually: keep if either != -1):")
print(df_wrong)
print()

# Correct: use & to require BOTH conditions
df_correct = df[(df.a != -1) & (df.b != -1)]
print("Using & (correct: keep only rows where both != -1):")
print(df_correct)
